curvefs/src/space/bitmap_allocator.cpp

```diff
@@ -0,0 +1,379 @@
+}
+
+struct BitmapAllocator::AllocOrder {
+    using Func = uint64_t (BitmapAllocator::*)(const uint64_t size,
+                                               const SpaceAllocateHint& hint,
+                                               Extents* exts);
+
+    AllocOrder(Func first, Func second, Func third)
+        : funs({first, second, third}) {}
+
+    const std::vector<Func> funs;
+};
+
+BitmapAllocator::BitmapAllocator(const BitmapAllocatorOption& opt)
+    : opt_(opt),
+      bitmapAreaLength_(CalcBitmapAreaLength(opt)),
+      smallAreaLength_(opt_.length - bitmapAreaLength_),
+      bitmapAreaOffset_(opt_.startOffset + smallAreaLength_),
+      mtx_(),
+      available_(opt_.length),
+      bitmapAllocIdx_(0),
+      bitmap_(bitmapAreaLength_ / opt_.sizePerBit),
+      bitmapExtent_(opt_.sizePerBit),
+      smallExtent_(opt_.startOffset, smallAreaLength_) {
+    assert(bitmap_.Size() * opt_.sizePerBit + smallExtent_.AvailableSize() ==
+           opt_.length);
+}
+
+BitmapAllocator::~BitmapAllocator() {}
+
+uint64_t BitmapAllocator::Alloc(uint64_t size, const SpaceAllocateHint& hint,
+                                Extents* exts) {
+    static const BitmapAllocator::AllocOrder kAllocBig(
+        &BitmapAllocator::AllocFromBitmap,
+        &BitmapAllocator::AllocFromBitmapExtent,
+        &BitmapAllocator::AllocFromSmallExtent);
+
+    static const BitmapAllocator::AllocOrder kAllocBigWithHint{
+        &BitmapAllocator::AllocFromBitmapExtent,
+        &BitmapAllocator::AllocFromBitmap,
+        &BitmapAllocator::AllocFromSmallExtent};
+
+    static const BitmapAllocator::AllocOrder kAllocSmall{
+        &BitmapAllocator::AllocFromSmallExtent,
+        &BitmapAllocator::AllocFromBitmapExtent,
+        &BitmapAllocator::AllocFromBitmap};
+
+    std::lock_guard<bthread::Mutex> lock(mtx_);
+    if (available_ == 0) {
+        return 0;
+    }
+
+    uint64_t alloc = 0;
+
+    if (size < opt_.sizePerBit && !IsAllocTypeBig(hint)) {
+        alloc = AllocInternal(kAllocSmall, size, hint, exts);
+    } else if (HasAllocOffsetHint(hint)) {
+        alloc = AllocInternal(kAllocBigWithHint, size, hint, exts);
+    } else {
+        alloc = AllocInternal(kAllocBig, size, hint, exts);
+    }
+
+    assert(alloc <= available_);
+    available_ -= alloc;
+
+    return alloc;
+}
+
+bool BitmapAllocator::DeAlloc(const uint64_t off, const uint64_t len) {
+    std::lock_guard<bthread::Mutex> lock(mtx_);
+
+    if (len > opt_.length - available_) {
+        return false;
+    }
+
+    uint64_t offInSmallExtent = 0;
+    uint64_t lenInSmallExtent = 0;
+    uint64_t offInBitmap = 0;
+    uint64_t lenInBitmap = 0;
+
+    Split(off, len, &offInSmallExtent, &lenInSmallExtent, &offInBitmap,
+          &lenInBitmap);
+
+    if (lenInSmallExtent != 0) {
+        DeAllocToSmallExtent(offInSmallExtent, lenInSmallExtent);
+    }
+
+    if (lenInBitmap != 0) {
+        DeAllocToBitmap(offInBitmap, lenInBitmap);
+    }
+
+    assert(available_ + len <= opt_.length);
```

Comment:
What if client dealloc twice? Here may assert.

Reply:
why need assert?  just return false and pirnt log info here.

Reply:
@cw123 currently, dealloc doesn't have this check. maybe we should check it later, but i think i''s client's duty to ensure only deallocate once.

@ilixiaocui Allocater has is own state, if this assert triggered, it means its state is broken, so it's better to make the process to quit. BTW, asserts here are just used for debug purpose, it will dissabled when NDEBUG macro defined.